eradiate/solvers/rami/app.py

```diff
@@ -185,13 +192,62 @@ def __attrs_post_init__(self):
                 })
 
     def kernel_dict(self, ref=True):
+        from eradiate.kernel.core import ScalarTransform4f
         result = KernelDict.empty()
 
+        # construct the dict for the combined surface and canopy;
+        # the BSDFs must be placed in the dict before the shapes,
+        # otherwise they can't be referenced
         if self.canopy is not None:
-            result.add(self.canopy)
+            canopy_dict = self.canopy.kernel_dict()
+
+            surface_dict = self.surface.kernel_dict()
+            bsdf_surface = surface_dict.pop("bsdf_surface")
+
+            patch_dict = {
+                "bsdf_surface": bsdf_surface,
+                "bsdf_leaf": canopy_dict["bsdf_leaf"],
+                "canopy": {
+                    **canopy_dict["canopy"],
+                    **surface_dict
+                }
+            }
+        else:
+            surface_dict = self.surface.kernel_dict()
+            bsdf_surface = surface_dict.pop("bsdf_surface")
+
+            patch_dict = {
+                "bsdf_surface": bsdf_surface,
+                "canopy": {
+                    "type": "shapegroup",
+                    "id": "canopy",
+                    **surface_dict
+                }
+            }
+
+        result.add(patch_dict)
+
+        # specify the instances for padding
+        kdu_length = kdu.get("length")
+        for x_offset in np.arange(-self.padding, self.padding+1):
+            for y_offset in np.arange(-self.padding, self.padding+1):
+                instance_dict = {
+                    "type": "instance",
+                    "group": {
+                        "type": "ref",
+                        "id": "canopy"
+                    },
+                    "to_world": ScalarTransform4f.translate([
+                            self.surface.width.m_as(kdu_length) * x_offset,
+                            self.surface.width.m_as(kdu_length) * y_offset,
+                            0.0
+                        ]
+                    )
+
+                }
+                result[f"instance{x_offset}_{y_offset}"] = instance_dict
```

Comment:
My idea was, that for the more complex canopy, e.g. the two layered ones, there would still only be one shapegroup with all leaves in it. But for that I think we will need nested shapegroups.

Reply:
I think we should ask RGL about nesting shapegroups. Maybe they tried and there are good reasons for not being able to nest them.

Reply:
Depends: are all trees identical or is each one unique?

Anyway, I think our real big problem is not with shape group nesting but instance nesting. I just tried and got a `RuntimeError: ​[ShapeGroup] Nested instancing is not permitted`. Means we'll need to see whether this can be solved at kernel level. I'll keep this as a discussion topic.

Reply:
Well yes... Kind of a shapegoup with an instance inside of it, right?

> Depends: are all trees identical or is each one unique?

At least in RAMI 3 there are at most two different kinds of trees in one canopy. But I don't think that is the main difference.
If you look at these ([one](https://rami-benchmark.jrc.ec.europa.eu/_www/phase/phase_exp.php?strTag=level4&strNext=meas&strPhase=RAMI3&strTagValue=HET_FLO_SOL_DIS); [two](https://rami-benchmark.jrc.ec.europa.eu/_www/phase/phase_exp.php?strTag=level2&strNext=meas&strPhase=RAMI3&strTagValue=HET_REA)) canopies, both would need nested instances. I suspect the number of leaves in each tree to be too large to not instance them.

Well it might be barely possible for those two. The 'real zoom in' scene with the two kinds of trees has about 13 million leaves in all of its 614 trees, which is barely possible on one machine with 16GB of ram

Reply:
BTW, having multiple groups with overlapping bounding boxes is likely less efficient.

Reply:
I understand what you mean, but how does this affect this problem? :thinking:

Reply:
> Kind of a shapegoup with an instance inside of it, right?
Yes. I don't thinkg this can be overcome easily, otherwise they'd have done it already.

> > having multiple groups with overlapping bounding boxes [...]
> how does this affect this problem? 🤔
It doesn't, that was just a though.

Reply:
As we talked about, I would like to add an abstraction for 'trees'. I will do that in a separate PR though.